### PR TITLE
GA4 Importer

### LINKED ISSUE #150: Non UA-... property IDs are not supported - support for  Google Analytics 4 Properties
Currently, property IDs that do not have the UA-* prefix result in an error when imported:

```
values must match the following regular expression: 'UA-[0-9]+-[0-9]+'"
```

This doesn't appear to originate from the plugin which only uses a regex to find the account ID if it has one, so must be coming from the SDK or API itself.

### PR DESCRIPTION
**Pull request by AltamashShaikh**
### Description:

Importer to import GA4 Data into Matomo.
Fixes: #150 
works with existing OAuth credentials
GA Data API and GA Admin API access is required
The GA Data and Admin API are still in beta phase

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Comment by AltamashShaikh** [comment]
Unsupported/Incompatible Doc
```
ga:exits, ga:entrances, ga:pageDownloadTime, ga:pageLoadSample, ga:goal{$gaIdGoal}Completions, ga:goal{$gaIdGoal}Value, ga:goal{$gaIdGoal}ConversionRate

ga:bounces(Bounces) is no more supported in GA4 but can be calculated using the bounceRate metric (bounceRate * sessions)

No Conditions for goals to fire, as goals are converted to events.
Couldn't find anything for segments in GA4 yet.

File: ImporterGA4.php [hunk 1]
```diff
@@ -0,0 +1,698 @@
...
+        try {
+            $extraCustomDimensions = $this->checkExtraCustomDimensions($extraCustomDimensions);
+
+            $webProperty = $this->gaAdminClient->getProperty($propertyId);
+
+            $startDate = Date::factory($webProperty->getCreateTime()->toDateTime()->getTimestamp())->toString();
+            if (!method_exists(SettingsServer::class, 'isMatomoForWordPress') || !SettingsServer::isMatomoForWordPress()) {
+                $idSite = SitesManagerAPI::getInstance()->addSite(
+                    $siteName = $webProperty->getDisplayName(),
+                    $urls = $type === \Piwik\Plugins\MobileAppMeasurable\Type::ID ? null : [$webProperty->getDisplayName()],
```
Comment (AltamashShaikh) [inline ImporterGA4.php:175]: API doesn't support any method to fetch the URL

File: ImporterGA4.php [hunk 2]
```diff
@@ -0,0 +1,698 @@
...
+                    $ecommerce = 1,
```
Comment (AltamashShaikh) [inline ImporterGA4.php:176]: No additional setting to fetch if its Ecommerce since its event based now, it supports ecommerce directly.

**Comment by Chardonneaur** [comment]
@AltamashShaikh  does it include GA row data? Could we get the log visit report thanks to it?

**Comment by AltamashShaikh** [comment]
@Chardonneaur I didn't understand the request sorry, can you give more details about it ?

**Comment by Chardonneaur** [comment]
@AltamashShaikh people claim that GA 4 allow you to access to everything, does it mean that we will get the log visit report filled with data with GA 4 importer?

**Comment by AltamashShaikh** [comment]
@Chardonneaur I will check if we can add it, but mostly that won't be possible as we migrate the aggregated data and not the granular data for GA Import.

**Comment by Chardonneaur** [comment]
@AltamashShaikh is GA4 blocking the export of granular data?

**Comment by AltamashShaikh** [comment]
@Chardonneaur There are row limits, [Refer](https://support.google.com/analytics/answer/10702008?hl=en)

**Comment by MarcoPNS** [comment]
Hey! Just wanted to test the current state of the GA4 Import. Is it correct that the import runs backwards?

Bildschirmfoto 2022-08-16 um 12 42 33

**Comment by AltamashShaikh** [comment]
@MarcoPNS  Yes the importer will try to import the most recent dates first, earlier it used to start from `start_date` and end with `end_date` now we try to import the recent dates first and today and other future dates at last

**Comment by justinvelluppillai** [comment]
@MarcoPNS this helps imports with a long date range as it can otherwise take days or weeks for the data to be available, so it is useful to have more recent data first. If this is a problem for your use case it'd be good to hear more about this to inform our future efforts.

**Comment by MarcoPNS** [comment]
I don't have a problem with it personally, but I don't think it's implemented properly yet, is it? Because currently, this does not work sooo great for me, unfortunately. The status does not update and new imports do not survive a rate limit. They change the status to "finished" because status and end-date is the same. 

So I changed the line to the previous for-loop so that I can import everything again:

https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/blob/d23b1b80355b887019ed88fdd6cbf981889d8e3f/ImporterGA4.php#L397-L398

**Comment by AltamashShaikh** [comment]
Hey @MarcoPNS can you give more detail on the issue you faced?
It will help us to replicate and fix the issue

**Comment by MarcoPNS** [comment]
Sure! I use Matomo 4.11.0 - I used the current live version of the GA-Importer. Then I removed it, cleared the cache and installed the 150-GA4-Import Branch manually. Then I tried to import a GA4 Property with following dates: 2021-08-16, 2022-08-16

Last imported date stays 2022-08-16 all the time. Last archieved date stays empty. After reaching the rate-limit and restarting the status changes to killed or finished.

Do such if-queries perhaps have something to do with it?
https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/blob/f261eff944593b869f620253171c5e3141d7a516/Importer.php#L437-L439

https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/blob/77c3f61b7e09890a7df6fe583fbba379caeb2184/Tasks.php#L182-L187

**Comment by AltamashShaikh** [comment]
@MarcoPNS I tried replicating but works fine for me may be the property I am testing doesn't have much data, can you please do this below steps to help us in debugging the issue ?

Revert the changes you did in [plugin-GoogleAnalyticsImporter/ImporterGA4.php](https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/blob/d23b1b80355b887019ed88fdd6cbf981889d8e3f/ImporterGA4.php#L397-L398)

Configure a new import with verbose logging enabled
Screenshot from 2022-08-18 13-45-44

There will be a log file created under /path/to/matomo/tmp/logs/ with a filename like `gaimportlog.{idSite_new_configured_account}..log`

Can you share that file ?

**Comment by MarcoPNS** [comment]
I left the import a little out of sight. When I looked again, the import was already "finished". 

Bildschirmfoto 2022-08-18 um 10 50 04

However, the import has again survived only till the rate limit.

Bildschirmfoto 2022-08-18 um 10 50 49

[gaimportlog.log](https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/files/9372021/gaimportlog.log)

**Comment by AltamashShaikh** [comment]
@MarcoPNS Thank you very much.
I can see we have an unhandled case for hourly limit.
Will add a fix and hopefully this should work

**Comment by AltamashShaikh** [comment]
@MarcoPNS Thanks for the log file, we just updated the code to handle the hourly rate limit exception.
This should not killed the import and eventually take alot more time since it will back of for an hour.

You need to update the plugin with newest change
Configure a new import and please configure a new import with verbose logging enabled so that we can check if its still breaking.

**Comment by MarcoPNS** [comment]
Hi! Its still breaking: [gaimportlog.log](https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/files/9393141/gaimportlog.log)

**Comment by AltamashShaikh** [comment]
@MarcoPNS I had added additional logging to know the status, but that also didn't worked.
I just updated the condition to catch the exception by `status` too.
Can we repeat the activity of updating the code and running a new import one more time ?

**Comment by MarcoPNS** [comment]
Here's the new log file:
[gaimportlog.log](https://github.com/matomo-org/plugin-GoogleAnalyticsImporter/files/9394705/gaimportlog.log)

**Comment by AltamashShaikh** [comment]
@MarcoPNS With the recent code changes, I can see its getting "rate_limited"
Screenshot from 2022-08-23 11-20-47

Are you sure you have update the correct code and branch ?

**Comment by AltamashShaikh** [comment]
@MarcoPNS Also updated the code to display "rate_limited_hourly" as a status instead of "rate_limited" for hourly rate limitation exception.
Screenshot from 2022-08-23 11-20-47-hourly

**Comment by AltamashShaikh** [comment]
@MarcoPNS After an hour the import ran and was finished successfully
Screenshot from 2022-08-23 12-32-03

**Comment by MarcoPNS** [comment]
Hi! I always disabled the plugin beforehand, cleared the cache, deleted the folder and then pulled and uploaded the latest version of your branch. Unfortunately, the status still jumps directly to finished although a lot of data is still missing.

If you need a good filled GA4 property to test this better maybe a [GA4 demo account](https://support.google.com/analytics/answer/6367342) will help.

**Comment by AltamashShaikh** [comment]
@MarcoPNS The demo account has no access to API unfortunately.
Getting this error.
Screenshot from 2022-08-24 13-48-13

**Comment by AltamashShaikh** [comment]
I tried hitting the limit for hour and I could see it working properly after the next hour.
I recently made few changes, you can update and check if its working. Also can you share what status do you see in the UI ?
Is it `killed`, `rate_limited` or `finished` ?

**Pull request merged**